File: app/src/main/java/ua/com/foxminded/locationtrackera/model/locations/LocationRepositoryImpl.java
package ua.com.foxminded.locationtrackera.model.locations;

import java.util.List;

import io.reactivex.rxjava3.core.Single;
import ua.com.foxminded.locationtrackera.model.locations.dao.LocationsDao;
import ua.com.foxminded.locationtrackera.model.locations.network.LocationsNetwork;
import ua.com.foxminded.locationtrackera.util.Result;

public class LocationRepositoryImpl implements LocationRepository {

    private final LocationsDao localDataSource;
    private final LocationsNetwork remoteDataSource;

    public LocationRepositoryImpl(LocationsDao localDataSource, LocationsNetwork remoteDataSource) {
        this.localDataSource = localDataSource;
        this.remoteDataSource = remoteDataSource;
    }

    @Override
    public void saveLocation(UserLocation userLocation) {
        localDataSource.saveLocation(userLocation);
    }

    @Override
    public List<UserLocation> getAllLocations() {
        return localDataSource.getAllLocations();
    }

    @Override
    public void deleteLocationsFromDb() {
        localDataSource.deleteAllLocation();
    }

    @Override
    public Single<Result<Void>> sendLocations(List<UserLocation> locationList) {
        return remoteDataSource.sendLocations(locationList);
    }

    @Override
    public Single<Result<List<UserLocation>>> retrieveLocations(double startDate, double endDate) {
        return remoteDataSource.retrieveLocations(startDate, endDate);
    }
}

File: models_impl/src/main/java/ua/com/foxminded/locationtrackera/models_impl/locations/LocationRepositoryImpl.kt
package ua.com.foxminded.locationtrackera.models_impl.locations

import io.reactivex.rxjava3.core.Single
import ua.com.foxminded.locationtrackera.models.locations.LocationRepository
import ua.com.foxminded.locationtrackera.models.locations.UserLocation
import ua.com.foxminded.locationtrackera.models.locations.dao.LocationsDao
import ua.com.foxminded.locationtrackera.models.locations.network.LocationsNetwork
import ua.com.foxminded.locationtrackera.models.util.Result

class LocationRepositoryImpl(
    private val localDataSource: LocationsDao,
    private val remoteDataSource: LocationsNetwork
) : LocationRepository {

    override fun saveLocation(userLocation: UserLocation) = localDataSource.saveLocation(userLocation)

    override fun deleteLocationsFromDb() = localDataSource.deleteAllLocation()

    override fun getAllLocations(): List<UserLocation> = localDataSource.getAllLocations()

    override fun sendLocations(
        locationList: List<UserLocation>
    ): Single<Result<Void?>> = remoteDataSource.sendLocations(locationList)

    override fun retrieveLocations(
        fromTime: Long, toTime: Long
    ): Single<Result<List<UserLocation>>> = remoteDataSource.retrieveLocations(fromTime, toTime)
}
